File: q68_longest_palindromic_substring/sliding_window.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        window = ""
        length = len(s)
        while length > 0 :
            window = s[:length]
            if window == window[::-1] :
                return window
            for i in range(length, len(s)) :
                window = window[1:] + s[i]
                if window == window[::-1] :
                    return window
            length -= 1

File: questions/q525_binary_subarrays_with_sum/sliding_window.cpp
class Solution {
public:
    int slidingWindowAtMostSum(vector<int>& nums, int goal) {
        if (goal < 0) {
            return 0;
        }

        int front=-1, rear=0;
        int curSum = 0;

        // Count measures the number of sub-arrays of length at least goal
        int count = 0;

        while (rear < nums.size()) {
            curSum += nums[rear];

            while (front <= rear && curSum > goal) {
                front++;
                curSum -= nums[front];
            }

            // Say the window is of size 3, we add the window length as we can form 3 sub-arrays starting from front.
            count += (rear - front);
            rear ++;
        }

        return count;
    }

    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return slidingWindowAtMostSum(nums, goal) - slidingWindowAtMostSum(nums, goal-1);
    }
};
